### PR TITLE
Support macos

### PR DESCRIPTION
**Pull request by isuruf**
Replaces https://github.com/flang-compiler/flang/pull/592

### CODE REVIEW

**Comment by d-parks** [comment]
Alternatively instead of removing the avx and sse versions of ceil/floor, you might want to update the dispatch tables in libpgmath/lib/x86_64/math_tables/mth*.h to use the SSE and AVX entry points so that SSE/AVX transition penalties can be avoided when running on AVX systems (although newer Intel processors do not suffer from large transition penalties).

**Comment by isuruf** [comment]
@d-parks, I've taken a stab in the dark. Please have a look.

**Comment by isuruf** [comment]
@d-parks, can you take a look at d699efa2482909e7afd42ac37e39b07b787c9e18?

**Comment by kiranchandramohan** [comment]
@isuruf Can you share the build instructions on Mac?

**Comment by isuruf** [comment]
@kiranchandramohan, it's the same as https://github.com/flang-compiler/flang/wiki/Building-Flang, but with `/usr/bin/gcc` replaced with `/usr/bin/clang`. You also need https://github.com/flang-compiler/classic-flang-llvm-project/pull/57

File: runtime/flangrti/CMakeLists.txt [hunk 1]
```diff
@@ -96,14 +96,14 @@ if (NOT DEFINED LIBOMP_EXPORT_DIR)
     FLANG_LIBOMP
     NAMES omp libomp
     HINTS ${CMAKE_BINARY_DIR}/lib)
-  target_link_libraries(flangrti_shared PRIVATE ${FLANG_LIBOMP})
+  target_link_libraries(flangrti_shared PUBLIC ${FLANG_LIBOMP})
```
Comment (pawosm-arm) [inline runtime/flangrti/CMakeLists.txt:99]: this commit deserves separate PR

**Review by pawosm-arm — CHANGES_REQUESTED** [review]

File: runtime/flangrti/CMakeLists.txt (hunk 1 above)
Comment (isuruf) [reply runtime/flangrti/CMakeLists.txt:99]: All 10 commits are independent and they all require separate PRs? I'd rather have one PR than 10 PRs which would take forever to review because of the lack of reviewers.

Comment (bryanpkc) [reply runtime/flangrti/CMakeLists.txt:99]: Changing the link dependencies of libraries has a much wider impact than the rest of the changes in this PR. Why is this change necessary?

Comment (isuruf) [reply runtime/flangrti/CMakeLists.txt:99]: It's described in the commit message.

>  libomp and libpgmath are dependencies of libflangrti's downstream libflang

> Since flang runtime library libflang has fortran files and they depend on
libomp and libpgmath, they should be public

**Review by bryanpkc — APPROVED** [review]
LGTM.

**Comment by isuruf** [comment]
Done

**Review by bryanpkc — APPROVED** [review]
LGTM.

File: runtime/flangrti/ktrap.c [hunk 2]
```diff
@@ -50,7 +50,11 @@ __ktrap(void)
         excepts |= FE_UNDERFLOW;
       if (bv & 0x100)
         excepts |= FE_INEXACT;
+#ifdef TARGET_OSX
```
Comment (pawosm-arm) [inline runtime/flangrti/ktrap.c:53]: Tried this on my mac. Neither of `__fenv_feenableexcept()` or `feenableexcept()` could be found...

**Comment by pawosm-arm** [comment]
I've tried to build classic flang on my mac for the first time. Some changes were needed:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 93dd9578..dbe363f7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -50,8 +50,8 @@ elseif (${TARGET_ARCHITECTURE} MATCHES "^(aarch64|arm64)$")
   set(TARGET_ARCHITECTURE aarch64)
   set(ARCHNAME aarch64)
   set(ARCH ARM)
-  add_definitions(-DTARGET_SUPPORTS_QUADFP)
-  set(TARGET_SUPPORTS_QUADFP True)
+#  add_definitions(-DTARGET_SUPPORTS_QUADFP)
+#  set(TARGET_SUPPORTS_QUADFP True)
 elseif( ${TARGET_ARCHITECTURE} STREQUAL "ppc64le" )
   set(ARCHNAME ppc64le)
   set(ARCH POWER)
@@ -292,7 +292,7 @@ if (LLVM_COMPILER_IS_GCC_COMPATIBLE)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -fno-common -Woverloaded-virtual -Wcast-qual -fno-strict-aliasing -pedantic -Wno-long-long -Wall -W -Wno-unused-parameter -Wwrite-strings")
   option(WITH_WERROR "Compile with '-Werror' C/C++ compiler flag" ON)
   if (WITH_WERROR)
-     set(CMAKE_C_FLAGS   "${CMAKE_C_FLAGS} -Werror -Wno-error=sign-compare -Wno-error=pointer-sign -Wno-error=incompatible-pointer-types -Wno-error=string-conversion -Wno-error=c99-extensions -Wno-error=deprecated-register -Wno-error=format -Wno-error=format-pedantic -Wno-error=tautological-constant-out-of-range-compare -Wno-error=switch -Wno-error=covered-switch-default -Wno-error=enum-compare -Wno-error=char-subscripts -Wno-error=parentheses -Wno-error=array-bounds -Wno-error=fortify-source -Wno-error=comment")
+     set(CMAKE_C_FLAGS   "${CMAKE_C_FLAGS} -Werror -Wno-error=deprecated-declarations -Wno-error=sign-compare -Wno-error=pointer-sign -Wno-error=incompatible-pointer-types -Wno-error=string-conversion -Wno-error=c99-extensions -Wno-error=deprecated-register -Wno-error=format -Wno-error=format-pedantic -Wno-error=tautological-constant-out-of-range-compare -Wno-error=switch -Wno-error=covered-switch-default -Wno-error=enum-compare -Wno-error=char-subscripts -Wno-error=parentheses -Wno-error=array-bounds -Wno-error=fortify-source -Wno-error=comment")
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror -Wno-error=sign-compare -Wno-error=pointer-sign -Wno-error=incompatible-pointer-types -Wno-error=string-conversion -Wno-error=c99-extensions -Wno-error=deprecated-register -Wno-error=format -Wno-error=format-pedantic -Wno-error=tautological-constant-out-of-range-compare -Wno-error=switch -Wno-error=covered-switch-default -Wno-error=enum-compare -Wno-error=char-subscripts -Wno-error=parentheses -Wno-error=array-bounds -Wno-error=fortify-source -Wno-error=comment")
      if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL 13)
        set(CMAKE_C_FLAGS   "${CMAKE_C_FLAGS} -Wno-error=unused-but-set-variable")
diff --git a/runtime/flangrti/ktrap.c b/runtime/flangrti/ktrap.c
index 573152d6..308604cd 100644
--- a/runtime/flangrti/ktrap.c
+++ b/runtime/flangrti/ktrap.c
@@ -9,7 +9,7 @@
  *  \brief IEEE trap support
  */
 
-#ifndef TARGET_WIN
+#if !defined(TARGET_WIN) && !defined(TARGET_OSX)
 
 #include <fenv.h>
 
diff --git a/runtime/flangrti/tempnam.c b/runtime/flangrti/tempnam.c
index 7c6602f7..90541640 100644
--- a/runtime/flangrti/tempnam.c
+++ b/runtime/flangrti/tempnam.c
@@ -5,7 +5,7 @@
  *
  */
 
-#if !defined(WINNT) && !defined(USETEMPNAM) /* { */
+#if !defined(WINNT) && !defined(USETEMPNAM) && !defined(TARGET_OSX) /* { */
 #include <errno.h>
 #include <stdio.h>
 
diff --git a/runtime/libpgmath/CMakeLists.txt b/runtime/libpgmath/CMakeLists.txt
index dde2f682..f651ec9a 100644
--- a/runtime/libpgmath/CMakeLists.txt
+++ b/runtime/libpgmath/CMakeLists.txt
@@ -61,8 +61,8 @@ if(${LIBPGMATH_SYSTEM_PROCESSOR} MATCHES "aarch64" AND NOT ${LIBPGMATH_WITH_GENE
   else()
     message(STATUS "Libpgmath CPU target explicitly set to ${LLVM_FLANG_CPU_TARGET}")
   endif()
-  add_definitions(-DTARGET_SUPPORTS_QUADFP)
-  set(TARGET_SUPPORTS_QUADFP True)
+#  add_definitions(-DTARGET_SUPPORTS_QUADFP)
+#  set(TARGET_SUPPORTS_QUADFP True)
 endif()
 
 if(CMAKE_C_COMPILER_ID STREQUAL "GNU" AND ${LIBPGMATH_SYSTEM_PROCESSOR} MATCHES "x86_64|aarch64")
diff --git a/tools/flang2/flang2exe/expatomics.cpp b/tools/flang2/flang2exe/expatomics.cpp
index 995284d7..cbe94451 100644
--- a/tools/flang2/flang2exe/expatomics.cpp
+++ b/tools/flang2/flang2exe/expatomics.cpp
@@ -1778,7 +1778,7 @@ get_ops(MSZ msz, int is_openmp)
   }
 }
 
-#if TARGET_LLVM_ATOMICS
+#if TARGET_LLVM_ATOMICS && !defined(TARGET_OSX)
 /** Given a size operand, return corresponding MSZ if operand is a constant.
     Otherwise return MSZ_UNDEF. */
 static MSZ
diff --git a/tools/flang2/flang2exe/llassem.cpp b/tools/flang2/flang2exe/llassem.cpp
index ee85e464..0893dc9f 100644
--- a/tools/flang2/flang2exe/llassem.cpp
+++ b/tools/flang2/flang2exe/llassem.cpp
@@ -175,12 +175,6 @@ static SPTR bss_base;
 static char bss_name[MXIDLN];
 static LL_ObjToDbgList *bss_dbg_list;
 static int ag_ctors_cnt = 0;
-#if defined(TARGET_OSX)
-static int emitted_bss_name = 0;
-static int emitted_static_name = 0;
-static int emitted_outer_bss_name = 0;
-static int emitted_outer_static_name = 0;
-#endif
 static char outer_static_name[MXIDLN]; /* Fortran: name of STATIC area for outer
                                           function */
 static char contained_static_name[MXIDLN]; /* Fortran: name of STATIC area for
```
Also, just in case I've applied this: https://github.com/flang-compiler/flang/pull/592/commits/834073192fcf9673fbd56340210ec7f4deeae5c4 and bunch of my PRs: #135 #546 and #1220

File: runtime/flangrti/ktrap.c (hunk 2 above)
Comment (isuruf) [reply runtime/flangrti/ktrap.c:53]: Looks like the function was moved from flangrti to libpgmath. https://github.com/flang-compiler/flang/blob/master/runtime/libpgmath/lib/common/fltfenv.c
